### ISSUE TITLE
[Request] Lint Testing Folder

### ISSUE THREAD

**Issue by weddi-eddy** [zmap/zlint#346]
Would it make sense to add an **experimentals** folder for lints such as this one that are suggested but need additional testing and comment? I guess by default, these should be excluded in the API, but can be included for CA and auditor development environments. It gives internal devs an easier to way to test it and possibly provide feedback.

If you did this however, you would need to think about how experimental lints are stored. If you had 100 suggested lints for testing, not everyone would want to download them in a production environment, so some thought would need to go into handling that.

I still think it's a good idea to consider. It gives a specific folder for certificate testers, CAs and potential lint writers to play in without messing with the official folders.  Maybe add a switch for the command to include lints in the **experimentals** folder during check.

(Also, thanks to @cpu for the response in pull 336 - Didn't want to update that pull for the thanks!)

**Comment by sleevi** [comment]
Setting aside the implementation for a second, what's the workflow you're thinking for how ZLint users will use this?

That is, it seems like as a production CA, you'd want to make sure you enable lints you're confident in: no false negatives and no false positives, ideally. You would not want to enable additional lints that had false positives (those could cause issuance rejection), and you wouldn't want to rely on lints that have false negatives (allowing misissued certificates)

So when would they enable these lints? When would they review these lints?

Figuring out how the ZLint user would use these, and what they would do in response, seems important to figuring out what/how to accomplish this.

For example, it might be that CAs enable the lints but suppress certain errors from certain lints from being blocking. But that assumes that every CA who does so will both contribute back (so that there are no more errors) and remember to disable the suppression. If, for example, two CAs both suppress the error, one sets a reminder to contribute in a month, the other sets a reminder to contribute in a year, then when the first CA contributes back and fixes whatever underlying issue in the lint, the second CA will be unprotected for 11 months because they've suppressed the lint.

Similarly, another idea might be for CAs to opt-in  to certain lints. Yet that seems unlikely, in practice, and unless the CAs that opt-in to a given lint issue certificates that are tested by that given lint, it doesn't seem likely to help. For example, a US CA that doesn't issue Qualified Certificates would never end up exercising the PSD2-related lints.

That's why it might be useful to work backwards, from trying to figure out how folks use a thing in practice :)

**Comment by weddi-eddy** [comment]
So, here's where the idea stemmed from.  The PSD2 new CABF OIDs don't have to be in place until the end of this month (don't quote me on that... but I believe that's the deadline). When new fields are implemented, we perform a ton of testing of the new fields in a non-trusted development environment.  When I send a certificate through ZLint (pre-certificate or otherwise), I get the results back and put them through my personal script which cleans it up, color codes, etc.

This isn't limited to PSD2, but it's the perfect example.  I would love to test these against our test implementations and certificates and make a judgement on whether the lint is sound, whether our changes are sound, etc.  It's an easier way of doing this (in my eyes).  There are also non-trusted CAs that although don't need to conform to the many various rules that trusted CAs need to, it's nice to be able to write custom lints and easily implement them for personal and private needs.  ZLint is a great tool to assist with conforming with rules and in conjunction with a CA.

Does that all make sense?

**Comment by zakird** [comment]
If this is the use case, then I think there are better solutions in typical development/release lifecycles. We can (and likely should) move to a solution where a RC is released ahead of time for something like that. Folks can run the RC in their development/testing environment and then we can release the major version after folks have had a chance to test. The proposed solutions mixes provenance and release cycle to me, which is awkward when we need to later go back and move the lints around.

**Comment by weddi-eddy** [comment]
Yeah! That's an acceptable approach and solves my issue.  I think that there needs to be a defined time between RC and release.  Compliance teams often ask when a certain lint is to be released and it's nice to have an answer (either based on date or time lapse of RC to R).

**Comment by cpu** [comment]
> We can (and likely should) move to a solution where a RC is released ahead of time for something like that.

I'm open to that :+1:

**Comment by weddi-eddy** [comment]
I know for our DEV systems, we automate some of the builds and upgrades (and keep a copy of existing versions and lints for backup purposes).  Would being able to obtain a copy of this RC be a trivial task through a Github command?  I apologize if this is a simple question. I'm fairly new to github and it pull procedures.

**Comment by cpu** [comment]
>  Would being able to obtain a copy of this RC be a trivial task through a Github command?

It should be possible. There is a way to work with "releases" [through the Github API](https://developer.github.com/v3/repos/releases/). One of the fields specified for releases is a `prerelease` bool. Likely we will publish RC releases to Github with this bool set to true.

**Comment by cpu** [comment]
I opened an issue to track the discussion on providing RCs: https://github.com/zmap/zlint/issues/351

@weddi-eddy If you agree that would meet the needs expressed in this issue I'd appreciate if you could close it out so we can keep things tidy. Thanks!

**Comment by weddi-eddy** [comment]
Thanks @cpu! You guys rock!
